### ISSUE TITLE
[Firefox Android] not working post 1.13.8

### ISSUE THREAD

**Issue by nyvanga** [gorhill/uBlock#2962]
When ublock0 first auto upgraded to 1.13.10 it was disabled in Firefox Android. Since then I tried various beta versions (1.14.*) and lastly 1.14.4. The problems ranges from Firefox hanging completely, to ublock not working to 1.14.4 where ublock seems to work but not filter anything. After reverting to 1.13.8 it seems fine.

**Comment by gorhill** [comment]
> seems to work but not filter anything

How did you come to the conclusion that it does not filter anything?

There are UI issues with Firefox for Android 55, which are fixed with Firefox for Android 56 and Nightly. It is unfortunate that I had to push webext version of uBO earlier than planned, but the webext-hybrid version was causing too much installation issues all over the place.

See my notes regarding [Firefox for Android in release notes of 1.14.6](https://github.com/gorhill/uBlock/releases/tag/1.14.6):

> There are issues with accessible uBO's dashboard or popup panel in Firefox for Android 55. These issues are resolved in Firefox for Android Nightly, and also in Firefox for Android Beta I believe. The WebExtensions version of uBO was not supposed to ship before Firefox 56 was released, but due to the major installtion issue affecting the webext-hybrid version of uBO, I had no choice to publish the pure webext version. One of the consequence is that Firefox for Android webext framework has issues in dealing with extensions' UI.
> 
> See issue #2958 for discussion.

**Comment by nyvanga** [comment]
> How did you come to the conclusion that it does not filter anything?
I tested a Danish news site (bt.dk) and there was no counter on the menu + it seemed to me that banners was shown (at least the "this site uses cookies"), which usually are not shown. So no science, just visual and a feeling.

**Comment by gorhill** [comment]
> there was no counter on the menu

uBO uses `browserAction.setTitle` to update the count: <https://bugzilla.mozilla.org/show_bug.cgi?id=1351111>.

But yes, I do observe that it does not seem to be properly updated with Firefox for Android 55, while I see it working fine with Firefox for Android Nightly.

Could be <https://bugzilla.mozilla.org/show_bug.cgi?id=1386786> maybe? It's marked as fixed in Firefox for Android 56.

**Comment by gorhill** [comment]
Regarding the cookie notice: maybe related to the fact that uBO's setting _"Ignore generic cosmetic filters"_ is checked by default on Firefox for Android?

**Comment by nyvanga** [comment]
Well I don't know about that, but 13.8 seems to remove the banner.
Have now tried 14.4 with nightly Firefox, and that was a lot smoother experience. So I will use 1.13.8 until Firefox 56 is released.

**Comment by superlex** [comment]
Hi,
uBlock Origin 1.14.* stable doesn't work for me when I open Firefox for Android 55.0.2 (but 56 beta too) because it doesn't block any ads but I noticed it start to work after a few of minutes that I'm using Firefox. I don't know why, however uBlock 1.13.8 works well. Maybe @nyvanga you could wait to see if you have the same problem.

**Comment by nyvanga** [comment]
Just ran another test with Firefox Nightly 57.0a1 (how do you test 56.*?) and ublock0 1.14.4, and it seems to block as much as 1.13.8 on 55.0.2.
I suspect that the reasons 1.14.* doesn't seem to block anything on 55.0.2, is that it is impossible to get into the dashboard. And since I usually have all blocking lists enabled (except country specific ones), the default setup probably blocks a lot less than I am used to.

**Comment by nyvanga** [comment]
Finally found Firefox beta 56.0b9 and tried it with 1.14.4. And can confirm it blocks less. It is weird though. It reports the same block count (20) but my infamous cookie banner stays. It is removed both by 55.0.2 + 1.13.8 and 57.0a1 + 1.14.4.
(All 3rd-party filters enabled down to "Regions, languages" and also "DNK Schacks Adblock Plus Liste". All other settings left at default.)

**Comment by jjohns71** [comment]
I think there is a Timing/Load/Cache issue with uBlock Origin on Firefox for Android - this is _probably_ WebExtenstions related and **NOT** uBO since the timing was fine with FFX android v<53.  The following is based on my **stupid user** observations.

Steps to ***always*** reproduce timing issue on my phone (Nexus 6P on Android 7.1.2 **_and_** Android 8.0.0) (using **both** FFX 55.0.4 and 56.0b9) are below.

1. From Firefox hamburger menu, choose 'Quit'. This will clear your personal data (if selected) and kill the browser completely.  
2. Now "cold boot" Firefox by launching it by clicking a web link in any other app. Example: [nytimes.com](https://www.nytimes.com/)
3. Firefox will now launch
4. Observe that uBO apparently is "behind" the FFX launch by clicking on the FFX hamburger menu and navigating to 'uBlock Origin' 
5. From uBO popup window, click on the element picker tool and observe that the **element picker tool will not work.**
6. Copy and paste the web address and paste it into a **new** tab after uBO has been able to load
7. Go to the UBO menu and click on the element picker tool - The **element picker tool will not work.**
8. uBO works as it should

tl:dr: Firefox WebExtensions rollout not good.
@gorhill

**Comment by nyvanga** [comment]
I have observed something similar. And I always quit Firefox and clear most data.
Usually I wait for everything to settle, before accessing the first website. Once in a while when I forget that, it seems that the website loading was delayed until everything was ready. (Again just "stupid user observations"). But I have noticed that if you cold boot FF 55 and click a bookmark immediately (bookmarks is my start page), it will load instantly if UBO is not installed, and halt for quite a while if it IS installed.
I just tested with FF 56.0b9 + 1.14.4 and a new tab actually worked. Then that annoying cookie popup is removed.

**Comment by gorhill** [comment]
> And I always quit Firefox and clear most data.

You mean cookies and site data? uBO is using indexedDB to store filter lists and their pre-parsed counterparts, so I wonder if uBO's indexedDB is affected. That would explain the longer delay before uBO is ready, downloading/parsing the filter lists is CPU intensive, hence why uBO cache the result of parsing them. See <https://bugzilla.mozilla.org/show_bug.cgi?id=1313401>.

**Comment by nyvanga** [comment]
I clear everything except "Saved logins". But I don't think it reloads everything. Or rather when I click refresh in the dashboard, it takes several minutes to complete. And the delay on startup is only a few seconds.

**Comment by gorhill** [comment]
> I click refresh in the dashboard, it takes several minutes to complete

You mean when you _"Purge all caches"_ then _"Update"_?

**Comment by gorhill** [comment]
> the delay on startup is only a few seconds

Well if it's only a few seconds, it's not unexpected. uBO has to load into memory tens of thousands of filters, so of course this will never be instantaneous. Typically, uBO if left to run normally will load from a sort of memory snapshot (I call this a _sekfie_), and at this point there is really nothing more I can do to improve load time.

**Comment by nyvanga** [comment]
Something like that yes. I usually only update when I choose lists. And that is only on first setup, so that would be like a purge + update.

**Comment by gorhill** [comment]
Ok let's go back to what the real issue was here, I am getting lost. The issue is titled _"not working post 1.13.8"_. So is it working or not? The load time really should not have changed that much with WebExtensions API. Now I find difficult to understand what is the issue after you said _"startup is only a few seconds"_, which means it _is_ working, which contradicts the title of the issue here.

**Comment by nyvanga** [comment]
Ok. Sorry for the confusion. Here is what I have found so far:
1. 55.0.2 + 1.13.8 works fine but anything after 1.13.8 does not. 
2. 56.0b9 + 1.14.4 does not block everything in the first tab, but works fine in subsequent tabs. 
3. 57.0a1 + 1.14.4 works fine.

**Comment by jjohns71** [comment]
I have not tested on nightly, but confirm that FFX Android 55.02 & uBO 1.13.8 were working fine with no boot loading issue.

After uBO migration to webextension, the uBO timing/cache delay on FFX startup is extremely noticeable on both FFX for Android 55.03 and FFX Android Beta 56.0b9 and reproducible using [the steps I described above](https://github.com/gorhill/uBlock/issues/2962#issuecomment-328007593).

**Comment by nyvanga** [comment]
I was able to get into the dashboard of UBO 1.14.4 on FF 55.0.2. I think it was because I loaded a local (on the phone) html file. Anyway I was now able to setup UBO 1.14.4 with all my usual settings. And I can now confirm that it blocks similar to 56.0b9 + 1.14.4. That is the first tab opened doesn't seem to block everything, but if a new tab is opened everything is blocked as it should.

**Comment by nyvanga** [comment]
I now use the following workaround to use Firefox 55.0.2 with ublock0 1.14.4:
1. Cold boot Firefox
2. Click on the tabs icon
3. Close the only open tab.
This will automatically launch a new tab, where ublock works and the dashboard is accessible.
